catchup/peerSelector.go

```diff
@@ -0,0 +1,287 @@
+		return httpPeer.GetAddress()
+	} else if unicastPeer, ok := peer.(network.UnicastPeer); ok {
+		return unicastPeer.GetAddress()
+	}
+	return ""
+}
+
+// refreshAvailablePeers reload the available peers from the network package, add new peers along with their
+// corresponding initial rank, and deletes peers that have been dropped by the network package.
+func (ps *peerSelector) refreshAvailablePeers() {
+	existingPeers := make(map[string]network.Peer)
+	for _, pool := range ps.pools {
+		for _, localPeer := range pool.peers {
+			if peerAddress := peerAddress(localPeer.peer); peerAddress != "" {
+				existingPeers[peerAddress] = localPeer.peer
+			}
+		}
+	}
+	sortNeeded := false
+	for _, initClass := range ps.peerClasses {
+		peers := ps.net.GetPeers(initClass.peerClass)
+		for _, peer := range peers {
+			peerAddress := peerAddress(peer)
+			if peerAddress == "" {
+				continue
+			}
+			if _, has := existingPeers[peerAddress]; has {
+				delete(existingPeers, peerAddress)
+				continue
+			}
+			// it's an entry which we did not had before.
+			sortNeeded = ps.addToPool(peer, initClass.initialRank, initClass) || sortNeeded
+		}
+	}
+
+	// delete from the pools array the peers that do not exist on the network anymore.
+	for poolIdx := len(ps.pools) - 1; poolIdx >= 0; poolIdx-- {
+		pool := ps.pools[poolIdx]
+		for peerIdx := len(pool.peers) - 1; peerIdx >= 0; peerIdx-- {
+			peer := pool.peers[peerIdx].peer
+			if peerAddress := peerAddress(peer); peerAddress != "" {
+				if _, has := existingPeers[peerAddress]; has {
+					// need to be removed.
+					pool.peers = append(pool.peers[:peerIdx], pool.peers[peerIdx+1:]...)
+				}
+			}
+		}
+		if len(pool.peers) == 0 {
+			ps.pools = append(ps.pools[:poolIdx], ps.pools[poolIdx+1:]...)
+			sortNeeded = true
+		} else {
+			ps.pools[poolIdx] = pool
+		}
+	}
+
+	if sortNeeded {
+		ps.sort()
+	}
+}
+
+// findPeer look into the peer pool and find the given peer.
+// The method returns the pool and peer indices if a peer was found, or (-1, -1) otherwise.
+func (ps *peerSelector) findPeer(peer network.Peer) (poolIdx, peerIdx int) {
+	peerAddr := peerAddress(peer)
+	if peerAddr != "" {
+		for i, pool := range ps.pools {
+			for j, localPeerEntry := range pool.peers {
+				if peerAddress(localPeerEntry.peer) == peerAddr {
+					return i, j
+				}
+			}
+		}
+	}
+	return -1, -1
+}
+
+// calculate the duration rank by mapping the range of [minDownloadDuration..maxDownloadDuration] into the rank range of [minRank..maxRank]
+func downloadDurationToRank(downloadDuration, minDownloadDuration, maxDownloadDuration time.Duration, minRank, maxRank int) (rank int) {
+	if downloadDuration < minDownloadDuration {
+		downloadDuration = minDownloadDuration
+	} else if downloadDuration > maxDownloadDuration {
+		downloadDuration = maxDownloadDuration
+	}
+	rank = minRank + int((downloadDuration-minDownloadDuration).Nanoseconds()*int64(maxRank-minRank)/(maxDownloadDuration-minDownloadDuration).Nanoseconds())
```

Comment:
Can you please add some explanation on this formula?

Reply:
will do.